### ISSUE TITLE
[`flake8-type-checking`] False positives/negatives for TC001-004 for overlapping module imports

### ISSUE THREAD

**Issue by Daverball** [astral-sh/ruff#15723]
### Description

This bug was discovered in #15719

Given something like:

```
from __future__ import annotations

import importlib.abc
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    import importlib.machinery

class Foo(importlib.abc.MetaPathFinder):
    def bar(self) -> importlib.machinery.ModuleSpec: ...
```

`importlib.machinery` will incorrectly be flagged with `TC004`, since `importlib.abc.MetaPathFinder` binds to the same binding as `importlib.machinery.ModuleSpec`, so since the former is used at runtime, it thinks that `importlib.machinery` needs to be moved outside the `TYPE_CHECKING` block.

Combining the two imports into a shared binding makes sense, since what will be looked up is `importlib` and not `importlib.machinery` or `importlib.abc`.

This means that when we're checking the import bindings we need to expand the single `importlib` binding to its two subbindings `importlib.abc` and `importlib.machinery` and for each reference check which of the two bindings will actually be relevant.

**Comment by coredumperror** [comment]
Could this possibly be why I'm getting `TC004` from the following code? 

```python
from typing import TYPE_CHECKING
from django.urls import reverse_lazy

if TYPE_CHECKING:
    from django_stubs_ext import StrOrPromise


class ApplicationTable:

    #: The URL to submit form actions to
    form_url: StrOrPromise = reverse_lazy("operations:applications--actions")
```

https://play.ruff.rs/d3e8ee0f-a7ba-477c-964c-d59bda9c396d

It doesn't seem to make any sense that this would be triggering TC004.

**Comment by Daverball** [comment]
No, it makes perfect sense that this would emit `TC004`.

It's because your target version is Python 3.9 and you're not using `from __future__ import annotations`, so all annotations are evaluated at runtime, so this code will fail with a `NameError` at runtime if you run it with Python 3.9 - 3.13, it only works like this on Python 3.14.

Overlapping imports are only a concern if you import entire submodules from the same parent package.
